advcpmv-0.9-8.32.patch

```diff
@@ -0,0 +1,702 @@
+diff -aur coreutils-8.32/src/copy.c coreutils-8.32-patched/src/copy.c
+--- coreutils-8.32/src/copy.c	2020-01-01 15:13:12.000000000 +0100
++++ coreutils-8.32-patched/src/copy.c	2021-02-16 23:24:31.301539409 +0100
+@@ -129,6 +129,157 @@
+   dev_t dev;
+ };
+ 
++/* BEGIN progress mod */
++
++struct progress_status {
++  int iCountDown;
++  char ** cProgressField;
++  struct timeval last_time;
++  int last_size, iBarLength;
++  struct stat src_open_sb;
++};
++
++/* Big Thanks to Sebastian from Stackoverflow for this code!
++ * https://stackoverflow.com/questions/1121383/counting-the-number-of-files-in-a-directory-using-c 
++ * */
++int countfiles(const char *path) {
++    DIR *dir_ptr = NULL;
++    struct dirent *direntp;
++    char *npath;
++    if (!path) return 0;
++    if( (dir_ptr = opendir(path)) == NULL ) return 0;
++
++    int count=0;
++    while( (direntp = readdir(dir_ptr)))
++    {
++        if (strcmp(direntp->d_name,".")==0 ||
++            strcmp(direntp->d_name,"..")==0) continue;
++        switch (direntp->d_type) {
++            case DT_REG:
++                ++count;
++                break;
++            case DT_DIR:            
++                npath=malloc(strlen(path)+strlen(direntp->d_name)+2);
++                count += countfiles(npath);
++                free(npath);
++                break;
++        }
++    }
++    closedir(dir_ptr);
++    return count;
++}
++
++/* Big Thanks to himero from www.unixboard.de for this code!
++ * https://www.unixboard.de/threads/verzeichnisgroesse-ermitteln.23595/
++ * Please note that it was a bit modified to test single files, too.
++ * */
++long getDirSize(const char *dirname) {
++    DIR *dir;
++    char fullDirName[1024];
++    struct dirent *dirPnt;
++    struct stat aktFile;
++    long long size = 0;
++
++    /* test single file */
++    sprintf(fullDirName, "%s", dirname);
++    if(lstat(fullDirName, &aktFile) != -1){
++        if(aktFile.st_mode & S_IFREG){
++            size += aktFile.st_size;
++        }
++    }
++
++    /*open the akt dir*/
++    sprintf(fullDirName, "%s/", dirname);
++    if((dir = opendir(fullDirName)) == NULL){
++        if (size > 0) {
++            return size / 1000;
++        } else {
++            return 0;
++        }  
++    }
++   
++    /*go through the directory*/
++    while( (dirPnt = readdir(dir)) != NULL ){
++        if(strcmp((*dirPnt).d_name, "..") == 0 ||
++            strcmp((*dirPnt).d_name, ".") == 0){
++            continue;
++        }
++   
++        /*create the akt filepath*/
++        sprintf(fullDirName, "%s/%s", dirname, (*dirPnt).d_name);
++        if(stat(fullDirName, &aktFile) == -1){
```

Comment:
stat() stats the target file. That's a problem.

I do not think we can just take random sample code from the internet to replace `du`/`find`. If you are using a OS where you don't have `du` and `find`, it's time to switch.

Reply:
The rest of the changes look fine.

Reply:
How can it stat the target file, if the function is given the source path (directory or file) to work on?

Don't you think code posted on stackoverflow or some open forum is meant to be used by others? Even this whole program itself is taken from the internet (the wayback macine even) and applied on the sourcecode of the coreutils which is taken from the internet, too?

Code on stackoverflow is released under CC BY-SA so it is free to use and modify even commercially - while this project isn't even that and open source too.

For the code from the open forum I don't know. I could try to contact the original owner. The post is from 2006 however, so I doubt he is still active in that forum.

Also both snippets are not "rocket science" but how everybody would need to tackle the problems if he wanted to do it as basic as possible, as there are no stock c functions for it.

The reason for replacing du and find is not that they are missing on my system. It is that calling them "externally" does not work, if the source file or directory name contains one or more double quotes. And I did encounter that quite often. It's not reassuring to copy or move data, when there are error messages that find and du did not work and then the global progress bar goes missing and the time counts overflow and don't work correctly, because the base values are missing or wrong. With find and du it looks like this (sorry for the german error messages - you could try on your system to get your localised ones):

![finduerror](https://user-images.githubusercontent.com/978967/108546828-f2b7a480-72e9-11eb-91ad-20ba60f2a442.png)

And it could look like this:
![finddureplaced](https://user-images.githubusercontent.com/978967/108547128-6063d080-72ea-11eb-8b60-3244ffe3eaaa.png)

I tried various different ways of escaping the calls to find and du differently but all of them failed before in the end deciding it would be easier to don't use them.

Reply:
> How can it stat the target file, if the function is given the source path (directory or file) to work on?

Please refer to `man stat`.

> It is that calling them "externally" does not work, if the source file or directory name contains one or more double quotes. And I did encounter that quite often.

@KlzXS can you take a look how to fix that?
The current code uses:

    sprintf ( command, "du -s '%s'", source );
    sprintf( command, "find \"%s\" -type f | wc -l", file[0]);
    sprintf ( command, "du -s \"%s\"", file[j] );

etc.

Reply:
I'll take a look at it tonight.

Reply:
I assume this is all for POSIX `sh`, right?

Reply:
Yes.

Reply:
Looking through man stat I noticed that it would be better to use stat instead of lstat so symbolic links are followed and the the size of the "target" file is returned. Was this what you meant? Then yes. You are correct. It is better to use stat instead of lstat. (I'm sorry for not noticing myself.)

@KizXS
What I already tried:

- was escaping in single quotes, instead of double quotes. (Will break if an single quote is in the filename).

- look if the source file/directory name contains a single or an double quote and used the opposite to escape. (Sadly this breaks if there are occurrences of both single and double quotes in the filename. Might not occur often, but if possible we should prevent such edge cases, too.)

A solution could be to escape each special char with a backslash, like e. g. bash does.

Reply:
> The current code uses:

Why are we mixing single and double quotes for arguments? Let's not do that, pick one.

A simple solution would be to replace every `'` with `'\''` or the double quote equivalent.

Reply:
@KizXS
The current code is, what is used before I did my changes and PR. Jarun is not fond of the idea, that I incorporated code I found on the net in my changes. He would like to keep using find and du.

You are right, that it would make sense, to use the same quotes for all calls to find and du. This will however not solve the problem, why I did remove the calls to find and du completely and replaced them with native c functions.

If a filename contains a quote, that is used in the quoted call to find and du, the call will not work. So if double quotes are used, a double quote in the filename will break the call.

If single quotes are used the call will break if there is a single quote in the filename. 

Detecting what quote type is in the filename and choosing the opposite one for the call programmatically will only work, if there are not both quote types used in the filename.

What Jarun would like us to do, as far as I understood, is for us to solve that problem and make the calls to find and du work with all possible filename combinations and not replace them. 

A solution could be to look through the source given by the user and escape all special chars with an backslash like the shells do on code completion. That way, no matter what quote is used for the call it work anyway.

Reply:
> @KizXS

Ah, the classic blunder. That's an L.

> This will however not solve the problem, why I did remove the calls to find and du completely and replaced them with native c functions.

That's not how I understood this situation. If you are using C functions there would be no need to escape anything as the C code would handle the strings as they are.

But if you are calling the shell I suggest to use the same quotes, as in not have both `'%s'` and `"%s"` to make the code simpler. Then, after choosing one, say `'`, you replace every occurance of it in the path string with `'\''`. This will properly escape both types of quotes when invoked in the shell.

So `SlrG's "file"` would become `SlrG'\''s "file"` and then you give that to the printf function. The shell calls `cmd 'SlrG'\''s "file"'` and does its thing.

Reply:
@KlzXS
Oh... Sorry. :( The first time I got it right. I don't know why I used an i afterwards.

Yes, thats the reason why I replaced the calls with C functions. Your suggestion is another good idea and will probably allow us to keep the find and du commands. I'll try to whip something up and if it works change my pull request accordingly. Thank you very much for sharing ideas.